File: Trabajo_3/Ejercicios/Ejercicio_cpp/Ejercicio_1.cpp
#include <iostream>
#include <vector>
using namespace std;

//Se declara la clase libro para evitar error por conexion con clase autor
class Libro;


//01
class Autor {
private:
    string nombre;
    string nacionalidad;
    vector<Libro*> librosEscritos;

public:
    //Constructor
    Autor(string name = "Grossman", string nacionality = "Peruano"){
        this -> nombre = name;
        this -> nacionalidad = nacionality;
    }

    //Getter, Setters
    string get_nombre() {
        return nombre;
    }

    string get_nacionalidad() {
        return nacionalidad;
    }

    void agregar_libro(Libro* libro) {
        librosEscritos.push_back(libro);
    }

    //Tanto esta como la de la clase usuarios nos servira para retornar los vectores y poder imprimirlos en el main.
    vector<Libro*> get_lista(){
        return librosEscritos;
    }
};

//02
class Usuario {
private:
    string nombre;
    string direccion;
    string telefono;
    vector<Libro*> librosPrestados;

public:
    Usuario(string name = "Joaquin", string direction = "Cayma", string cell = "123"){
        this -> nombre = name;
        this -> direccion = direction;
        this -> telefono = cell;
    }

    string get_nombre() {
        return nombre;
    }

    string get_direccion() {
        return direccion;
    }

    string get_telefono() {
        return telefono;
    }

    //Sirve para tener un vector de libros prestados por usuario
    void agregarLibroPrestado(Libro* libro) {
        librosPrestados.push_back(libro);
    }

    vector<Libro*> get_lista(){
        return librosPrestados;
    }

};

class Prestamo {
private:
    Usuario usuario;
    Libro* libro;
    string F_Prestamo;
    string F_Devolucion;

public:
    Prestamo(Usuario user, Libro* book, string Prestamo_F = "29/12/2005", string Devolucion_F = "13/04/2024"){
        this -> usuario = user;
        this -> libro = book;
        this -> F_Prestamo = Prestamo_F;
        this -> F_Devolucion = Devolucion_F;
    }
};


class Libro {
private:
    string ISBN;
    string titulo;
    string fecha_publi;
    int stock;
    Autor* autor;

public:
    Libro(string IS = "01", string title = "Hola", string date = "29/12/2005", int cant = 0, Autor* author = nullptr) {
        this -> ISBN = IS;
        this -> titulo = title;
        this -> fecha_publi = date;
        this -> stock = cant;
        this -> autor = author;
    }

    string get_ISBN() {
        return ISBN;
    }

    string get_titulo() {
        return titulo;
    }

    string get_AñoPublicacion() {
        return fecha_publi;
    }

    int get_Cantidad() {
        return stock;
    }

    Autor* get_Autor() {
        return autor;
    }
};

int main() {
    // Creamos autores:
    Autor Cervantes("Miguel de Cervantes", "Español");
    Autor Angel("Angel Zorrilla", "Haitiano");

    // Creamos libritos:
    Libro Pinocho("1234567890", "Pinocho", "-154", 18, &Cervantes);
    Libro Caperucita("0987654321", "Caperucita roja", "1232", 58, &Angel);

    // Agregar libros a los autores:
    Cervantes.agregar_libro(&Pinocho);
    Angel.agregar_libro(&Caperucita);

    // Creamos usuarios:
    Usuario Gras("Grossman", "Wa", "1234567890");
    Usuario Pedro("Pedro Pedro Pedro Pedro pe", "URSS", "0987654321");

    // Agregamos libros a los usuarios:
    Gras.agregarLibroPrestado(&Pinocho);
    Pedro.agregarLibroPrestado(&Caperucita);

    // Creamos prestamos:
    Prestamo prestamo_1(Gras, &Pinocho, "28/10/2000", "29/12/2020");
    Prestamo prestamo_2(Pedro, &Caperucita,"25/11/1995","14/05/1111");

    // Asignamos a un vector los libros de los autores para posteriormente imprimirlos:
    vector<Libro*> librosCervantes = Cervantes.get_lista();
    vector<Libro*> librosAngel = Angel.get_lista();

    cout << "Libros escritos por " << Cervantes.get_nombre() << ":" << endl;
    for (Libro* libro : librosCervantes) {
        cout << "-" << libro->get_titulo() << endl;
    }

    cout << "Libros escritos por " << Angel.get_nombre() << ":" << endl;
    for (Libro* libro : librosAngel) {
        cout << "-" << libro->get_titulo() << endl;
    }

    // Asignamos a un vector los libros prestados por cada usuario para imprimirlos:
    vector<Libro*> librosGras = Gras.get_lista();
    vector<Libro*> librosPedro = Pedro.get_lista();

    cout << "Libros prestados a " << Gras.get_nombre() << ":" << endl;
    for (Libro* libro : librosGras) {
        cout << "- " << libro->get_titulo() << endl;
    }

    cout << "Libros prestados a " << Pedro.get_nombre() << ":" << endl;
    for (Libro* libro : librosPedro) {
        cout << "- " << libro->get_titulo() << endl;
    }
    
    return 0;
}

//pedro pedro pedro pedro pe

File: Trabajo_3/Ejercicios/Ejercicio_Python/Ejercicio1.py
from datetime import datetime

class Libro:
    def __init__(self, ISBN, titulo, anoPublicacion, cantidadEjemplares, autor):
        self.ISBN = ISBN
        self.titulo = titulo
        self.anoPublicacion = anoPublicacion
        self.cantidadEjemplares = cantidadEjemplares
        self.autor = autor

class Autor:
    def __init__(self, nombre, nacionalidad):
        self.nombre = nombre
        self.nacionalidad = nacionalidad
        self.librosEscritos = []

class Usuario:
    def __init__(self, nombre, direccion, telefono):
        self.nombre = nombre
        self.direccion = direccion
        self.telefono = telefono
        self.librosPrestados = []

class Prestamo:
    def __init__(self, usuario, libro, fechaPrestamo, fechaDevolucion):
        self.usuario = usuario
        self.libro = libro
        self.fechaPrestamo = fechaPrestamo
        self.fechaDevolucion = fechaDevolucion

class Bibliotecas:
    def __init__(self):
        self.libros = []
        self.autores = []
        self.usuarios = []
        self.prestamos = []

    def agregarLibro(self, libro):
        if libro not in self.libros:
            self.libros.append(libro)
            print("Libro agregado")

    def agregarAutor(self, autor):
        if autor not in self.autores:
            self.autores.append(autor)
            print("Autor agregado")

    def agregarUsuario(self, usuario):
        if usuario not in self.usuarios:
            self.usuarios.append(usuario)
            print("Usuario agregado")

    def agregarPrestamo(self, prestamo):
        if prestamo not in self.prestamos:
            self.prestamos.append(prestamo)
            print("Prestamo agregado")

if __name__ == "__main__":
    libro1 = Libro("1234567890", "El Quijote", 1605, 100, Autor("Miguel de Cervantes", "Español"))
    libro2 = Libro("0987654321", "Don Juan Tenorio", 1844, 50, Autor("José Zorrilla", "Español"))
    libro3 = Libro("1357924680", "La Celestina", 1499, 75, Autor("Fernando de Rojas", "Español"))

    autor1 = Autor("Miguel de Cervantes", "Español")

    usuario1 = Usuario("Juan", "Calle 123", "1234567890")
    usuario2 = Usuario("Pedro", "Calle 456", "0987654321")

    prestamo1 = Prestamo(usuario1, libro1, datetime.now(), datetime.now())
    prestamo2 = Prestamo(usuario2, libro2, datetime.now(), datetime.now())

    biblioteca = Bibliotecas()

    biblioteca.agregarLibro(libro1)
    biblioteca.agregarAutor(autor1)
    biblioteca.agregarUsuario(usuario1)
    biblioteca.agregarPrestamo(prestamo1)
    biblioteca.agregarLibro(libro2)
